core/src/main/scala/io/projectglow/sql/expressions/LogisticRegressionGwas.scala

```diff
@@ -173,48 +168,169 @@ class NewtonArguments(
 }
 
 case class NewtonResult(
-    args: NewtonArguments,
+    args: NewtonIterationsState,
     logLkhd: Double,
     nIter: Int,
     converged: Boolean,
     exploded: Boolean)
 
 trait LogitTest extends Serializable {
+  type FitState
   def resultSchema: StructType
+  def canReuseNullFit: Boolean
+  def fitNullModel(phenotypes: Array[Double], covariates: SparkDenseMatrix): FitState
+  def runTest(
+      x: DenseMatrix[Double],
+      y: DenseVector[Double],
+      nullModelFit: Option[FitState]): InternalRow
 }
 
-trait LogitTestWithNullModelFit extends LogitTest {
-  def runTest(nullFit: NewtonResult): InternalRow
+trait LogitTestWithoutNullFit extends LogitTest {
+  final override val canReuseNullFit: Boolean = false
+
+  final override def fitNullModel(
+      phenotypes: Array[Double],
+      covariates: SparkDenseMatrix): FitState = {
+    throw new IllegalArgumentException("Test does not support null model fit")
+  }
 }
 
-trait LogitTestWithNullAndFullModelFit extends LogitTest {
-  // fullFitOpt is None iff nullFit fails to converge
-  def runTest(nullFit: NewtonResult, fullFitOpt: Option[NewtonResult]): InternalRow
+trait LogitTestWithNullFit extends LogitTest {
+  final override val canReuseNullFit: Boolean = true
 }
 
-object LikelihoodRatioTest extends LogitTestWithNullAndFullModelFit {
-  override def resultSchema: StructType = Encoders.product[LikelihoodRatioTestStats].schema
-  override def runTest(nullFit: NewtonResult, fullFitOpt: Option[NewtonResult]): InternalRow = {
-    if (!nullFit.converged || !fullFitOpt.get.converged) {
+object LikelihoodRatioTest extends LogitTestWithNullFit {
+  override type FitState = NewtonResult
+  override def resultSchema: StructType = Encoders.product[LogitTestResults].schema
```

Comment:
I think it's probably a good idea to make the result schema part of the interface. If we support non-Wald confidence intervals, we'll need this. I will make it an `override val` instead of an `override def`.